pyproject.toml

```diff
@@ -0,0 +1,181 @@
+[build-system]
+requires = [
+    "setuptools",
+    "wheel",
+    # On Windows, use the CasADi vcpkg registry and CMake bundled from MSVC
+    "casadi>=3.6.0; platform_system!='Windows'",
+    "cmake; platform_system!='Windows'",
+    ]
+build-backend = "setuptools.build_meta"
+
+[project]
+name = "pybamm"
+version = "23.9rc0"
```

Comment:
Making this [`dynamic`](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#dynamic-metadata) will reduce the number of files in which we have to bump the version for a release -
```
dynamic = ["version"],
```

Reply:
There are a lot of commits in this PR haha but I have tried this. It did not work and some bits of debugging revealed that it was because the `build_meta` build-backend, unlike its legacy counterpart, does not install the optional dependencies first before finding the version as a dynamic attribute. This breaks the PyBaMM import (it cannot import `pybtex` for the Citations class, `anytree`, etc.) and therefore is unable to extract the version string from __version.

I would suggest against doing this for now until we can figure out a way to make the imports truly optional (perhaps by adding warnings that mark limited functionality for a class or feature if a module cannot be imported). This is one of four more issues that I will be opening after this PR is merged (actually three, since I opened #3462 today) to mark as subsequent improvements. The good thing is that we can also look into adding support for `setuptools-scm` now with suitable modifications to the `update_version.py` script.

Reply:
Ah, sounds good!